File: 0143-reorder-list/0143-reorder-list.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func reorderList(_ head: ListNode?) {
        //slow x1 and fast pointer x2
        //then from slow + 1, reverse and cut the first half off from the second
        //then have a reference both and do your pointer switching technique


        var slow = head
        var fast = head
        while let s = slow, let f = fast, let _ = fast?.next {
            slow = s.next
            fast = f.next?.next
        }

        var endOfFirstHalf = slow
        var startOfSecondHalf = slow?.next
        endOfFirstHalf?.next = nil

        var prev: ListNode? = nil
        var curr = startOfSecondHalf
        while let node = curr {
            var temp = node.next
            node.next = prev
            prev = node
            curr = temp
        }

        var first = head
        var second = prev

        while let node1 = first, let node2 = second {
            var temp = node1.next
            node1.next = second
            first = node1.next

            var temp2 = node2.next
            node2.next = temp
            second = node2.next
            first?.next = temp2
        }
    }
}

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        
        """
        Even vs Odd Length for Linked List

        1. Slow and fast pointer, this will give you a starting point for the 
        second half of the list which I can reverse
        2. Reverse the second half of the list
        3. Do arithmetic between first node in first half and reversed second half
        """

        #Step 1
        slow, fast = head, head
        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next
        print("True 1")
        
        #Step 2
        secondHalf = slow.next

        slow.next = None #Set the end of the first half of the linked list
        
        prev = None
        curr = secondHalf
        while curr != None:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        reversedSecondHalf = prev

        #Step3
        list1 = head
        list2 = reversedSecondHalf
        while list1 != None and list2 != None:
            temp1 = list1.next
            list1.next = list2
            list1 = temp1

            temp2 = list2.next
            list2.next = list1
            list2 = temp2